p2p/raintree/network.go

```diff
@@ -25,46 +25,71 @@ type rainTreeNetwork struct {
 
 	peersManager *peersManager
 
+	cleanupLayerEnabled    bool // debug config only
+	redundancyLayerEnabled bool
+
 	// TECHDEBT(drewsky): What should we use for de-duping messages within P2P?
 	mempool map[uint64]struct{} // TODO (drewsky) replace map implementation (can grow unbounded)
 }
 
-func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook) typesP2P.Network {
+func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook, cfg modules.P2PConfig) typesP2P.Network {
 	pm, err := newPeersManager(addr, addrBook)
 	if err != nil {
 		log.Println("[ERROR] Error initializing rainTreeNetwork peersManager: ", err)
 	}
 
 	n := &rainTreeNetwork{
-		selfAddr:     addr,
-		peersManager: pm,
-		mempool:      make(map[uint64]struct{}),
+		selfAddr:               addr,
+		peersManager:           pm,
+		redundancyLayerEnabled: cfg.GetUseRainTreeRedundancyLayer(),
+		cleanupLayerEnabled:    cfg.GetUseRainTreeCleanupLayer(),
+		mempool:                make(map[uint64]struct{}),
 	}
 
 	return typesP2P.Network(n)
 }
 
 func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
-	return n.networkBroadcastAtLevel(data, n.peersManager.getNetworkView().maxNumLevels, getNonce())
+	return n.networkBroadcastAtLevel(data, int32(n.peersManager.getNetworkView().maxNumLevels), getNonce())
 }
 
-func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level uint32, nonce uint64) error {
-	// This is handled either by the cleanup layer or redundancy layer
-	if level == 0 {
-		return nil
-	}
+func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level int32, nonce uint64) error {
 	msg := &typesP2P.RainTreeMessage{
 		Level: level,
 		Data:  data,
 		Nonce: nonce,
 	}
+
+	// Redundancy layer
+	if level == 0 {
+		if n.redundancyLayerEnabled {
+			level, msg.Level = n.redundancyLayer()
+		} else {
+			if err := n.demote(msg); err != nil {
+				log.Println("Error demoting self during RainTree message propagation: ", err)
+			}
+			return nil
+		}
+	}
+
+	targets := n.getTargetsAtLevel(uint32(level))
+
+	// Cleanup layer
+	if level == -1 {
```

Comment:
<img width="917" alt="Screenshot 2022-12-01 at 7 20 40 PM" src="https://user-images.githubusercontent.com/1892194/205207642-5fe5b50f-43bb-47d0-9716-d833fdc2bdf4.png">

I understand your source of confusion. I believe the code is right and the text can be made clearer. 

`level == -1 ` only when we call `demo(level=0)` which calls `internalBroadcast(-1)`

It's weird to describe because the way the code flows, it'll work out, so i would suggest trying is changing the code a bit such that.

1. if level == 0:
- If cleanup is enabled => do cleanup
- if redundancy is anebled => do redundancy
2. Do not call `internalBroadcast` if level == 0

If my intuition & understanding is correct, the tests should pass after this change.

Reply:
🤯 you guys grilled my code for way less than this, come on... 

You say that the code is right but how? If `level` is `uint32` (casted or not) how can the if in line 78 ever be truthy?
Overlooked? OK, but you didn't flag that? 🤔